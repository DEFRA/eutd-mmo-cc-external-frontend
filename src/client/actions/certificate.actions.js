import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};