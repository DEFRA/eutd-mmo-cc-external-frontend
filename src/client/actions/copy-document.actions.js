import { showFullPageError, apiCallFailed, isError, CLEAR_ERRORS } from './index';
export const CONFIRM_COPY_DOCUMENT            = 'confirm_copy_document';
export const UNAUTHORISED_COPY_DOCUMENT       = 'unauthorised_copy_document';
export const CLEAR_COPY_DOCUMENT              = 'clear_copy_document';
export const REMOVE_CONFIRM_COPY_CERTIFICATE  = 'remove_confirm_copy_certificate';
export const ADD_ERROR_COPY_CERTIFICATE       = 'add_error_copy_certificate';
export const REMOVE_ERROR_COPY_CERTIFICATE    = 'remove_error_copy_certificate';

export const submitCopyCertificate = (payload) => async (dispatch, _getState, {orchestrationApi}) => {
  const {documentNumber, copyDocumentAcknowledged, voidOriginal, excludeLandings, journey} = payload;
  const config = {headers: {documentnumber: documentNumber}};

  try {
    let ipAddress;
    if (voidOriginal) {
      const ip = await orchestrationApi.get('/client-ip');
      ipAddress = ip.data;
    }

    return await orchestrationApi.post('/confirm-copy-certificate', { copyDocumentAcknowledged, voidOriginal, excludeLandings, journey, ipAddress }, config)
    .then(res => {
      dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: {
        documentNumber: res.data.documentNumber,
        copyDocumentAcknowledged: res.data.copyDocumentAcknowledged,
        voidDocumentConfirm: res.data.voidOriginal
      }});
      return res.data.newDocumentNumber;
    });
  } catch (e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch({ type: UNAUTHORISED_COPY_DOCUMENT }); return;
    } else {
      dispatch(apiCallFailed(e.response));
    }
  }
};

export const checkCopyCertificate = (documentNumber) => async (dispatch, _getState, {orchestrationApi}) => {
  const config = {
    headers: {
      documentnumber: documentNumber
    }
  };

  try {
    const res = await orchestrationApi.get('/check-copy-certificate', config);
    const { canCopy } = res.data;
    if (!canCopy) {
      dispatch(unauthorised());
    }
  } catch(e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch(unauthorised());
      throw new Error(`Unauthorised access ${e.response.status}`);
    } else {
      dispatch(apiCallFailed(e.response));
      throw new Error(`An error has occurred ${e.response.status} (${e.response.statusText})`);
    }
  }

  function unauthorised() { return { type: UNAUTHORISED_COPY_DOCUMENT }; }
};

export const addCopyDocument = (data) => (dispatch) => dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: data });
export const clearErrors = () => (dispatch) =>  dispatch({ type: CLEAR_ERRORS });
export const clearCopyDocument = () => (dispatch) => dispatch({ type: CLEAR_COPY_DOCUMENT });
export const unauthorisedCopyDocument = () => (dispatch) => dispatch({ type: UNAUTHORISED_COPY_DOCUMENT });