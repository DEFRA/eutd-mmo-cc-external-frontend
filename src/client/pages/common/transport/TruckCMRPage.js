import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import HintText from '@govuk-react/hint-text';
import { isEmpty } from 'lodash';
import { withTranslation } from 'react-i18next';

import {
  Main,
  Header,
  MultiChoice,
  GridRow,
  GridCol
} from 'govuk-react';

import SelectRadio from '../../../components/elements/SelectRadio';
import PageTitle from '../../../components/PageTitle';

import {
  addTransportDetails,
  saveTruckCMR,
  dispatchClearErrors,
  dispatchApiCallFailed,
  getTransportDetails,
  clearTransportDetails
} from '../../../actions';

import {
  getLandingType
} from '../../../actions/landingsType.actions';

import BackLinkWithErrorClearOut from '../../../components/elements/BackLinkWithErrorClearOut';
import { scrollToErrorIsland, onHandleErrorClick} from '../../utils';
import PageTemplateWrapper from '../../../components/PageTemplateWrapper';
import HelpLink from '../../../components/HelpLink';
import SaveAsDraftButton from '../../../components/SaveAsDraftButton';
import ContinueButton from '../../../components/elements/ContinueButton';
import ErrorIsland from '../../../components/elements/ErrorIsland';
import { catchCertificateJourney } from '../../../helpers/journeyConfiguration';
import BackToProgressLink from '../../../components/BackToProgressLink';
class TruckCMRPage extends Component {
  constructor(props) {
    super(props);
  }

  onChange = e => {
    this.props.addTransportDetails({ [e.target.name]: e.target.value });
  };

  onContinue = async e => {
    e.preventDefault();
    try {
      const { summaryUri, truckDetailsUri, path, journey } = this.props.route;
      const documentNumber = this.props.match.params['documentNumber'];
      const transport = await this.props.saveTruckCMR(path, journey, false, documentNumber);
      if (transport.cmr === 'true') {
        this.props.history.push(summaryUri.replace(':documentNumber', documentNumber));
      } else {
        this.props.history.push(truckDetailsUri.replace(':documentNumber', documentNumber));
      }
    } catch (error) {
      scrollToErrorIsland();
    }
  };

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    try {
      const { path, journey, saveAsDraftUri } = this.props.route;
      const documentNumber = this.props.match.params['documentNumber'];

      const isTruckCMRSavedAsDraft = true;
      await this.props.saveTruckCMR(path, journey, isTruckCMRSavedAsDraft, documentNumber);

      this.props.history.push(saveAsDraftUri);
    } catch (error) {
      scrollToErrorIsland();
    }
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  displayDashboardOptions() {
    this.props.history.push(this.props.route.saveAsDraftUri);
  }

  displayLandingEntryOptions() {
    const documentNumber = this.props.match.params['documentNumber'];
    const { landingsEntryUri } = this.props.route;
    this.props.history.push(landingsEntryUri.replace(':documentNumber', documentNumber));
  }

  hasLandingEntryOption() {
    const { getAddedLandingsType } = this.props;
    return getAddedLandingsType && getAddedLandingsType.landingsEntryOption !== null && getAddedLandingsType.generatedByContent === false;
  }

  hasRequiredData() {
    const { transport } = this.props;
    return !isEmpty(transport);
  }

  isDirectLanding() {
    return !isEmpty(this.props.getAddedLandingsType) && this.props.getAddedLandingsType.landingsEntryOption === 'directLanding';
  }

  componentWillUnmount() {
    this.props.clearTransportDetails();
  }

  componentDidMount = async () => {
    const { journey, progressUri } = this.props.route;
    const documentNumber = this.props.match.params['documentNumber'];
    window.scrollTo(0, 0);

    try {
      await this.props.getTransportDetails(this.props.route.journey, documentNumber);

      // requiredData for the last page in the journey
      if (!this.hasRequiredData() || this.props.transport.vehicle != 'truck') {
        this.props.history.push(progressUri.replace(':documentNumber', documentNumber));
        return;
      }

      if (journey === catchCertificateJourney) {
        await this.props.getLandingType(documentNumber);

        if (!this.hasLandingEntryOption()) {
          this.displayLandingEntryOptions();
          return;
        }

        if (this.isDirectLanding()) {
          this.displayDashboardOptions();
          return;
        }
      }
    } catch (e) {
      console.error(e);
    }
  };

  componentDidUpdate() {
    const { transport } = this.props;

    if (transport.unauthorised === true) {
      this.unauthorised();
    }
  }

  render = () => {
    const documentNumber = this.props.match.params['documentNumber'];

    const { errors, cmrError } = this.props.errors;
    const {
      cmr,
      nationalityOfVehicle,
      registrationNumber,
      departurePlace,
      vehicle } = this.props.transport;
    const { dispatchClearErrors: _dispatchClearErrors, history, t } = this.props;

    const saveAsDraftFormActionUrl = '/orchestration/api/v1/transport/truck/cmr/saveAsDraft';

    const {
      title,
      previousUri,
      summaryUri,
      truckDetailsUri,
      journey,
      path,
      saveAsDraftUri,
      progressUri
    } = this.props.route;

    return (
      <Main>
        <PageTitle
          title={`${!isEmpty(errors) ? 'Error: ' : ''}${t('commonDoYouHaveaRoadTransportDocumentTitle')} - ${t(title)}`}
        />
        {errors && (
          <ErrorIsland
            errors={errors.map(err => ({
              message: t(err.text),
              key: err.targetName
            }))}
            onHandleErrorClick={onHandleErrorClick}
          />
        )}
        <GridRow>
          <GridCol>
            <BackLinkWithErrorClearOut
              backUri={previousUri.replace(':documentNumber', documentNumber)}
              clearErrors={_dispatchClearErrors}
              history={history}
              labeltext={t('commonBackLinkBackButtonLabel')}
            ></BackLinkWithErrorClearOut>
            <Header>{t('sdDoYouHaveaRoadTransportDocumentHeader')}</Header>
            <br />
            <HintText>{t('sdDoYouHaveaRoadTransportDocumentHint')}</HintText>
          </GridCol>
        </GridRow>
        <form
          action="/orchestration/api/v1/transport/truck/cmr"
          method="POST"
          onSubmit={(e) => e.preventDefault()}
        >
          <input type="hidden" name="summaryUri" value={summaryUri} />
          <input type="hidden" name="truckDetailsUri" value={truckDetailsUri} />
          <input type="hidden" name="journey" value={journey} />
          <input type="hidden" name="vehicle" value={vehicle} />
          <input
            type="hidden"
            name="nationalityOfVehicle"
            value={nationalityOfVehicle}
          />
          <input
            type="hidden"
            name="registrationNumber"
            value={registrationNumber}
          />
          <input type="hidden" name="departurePlace" value={departurePlace} />
          <input type="hidden" name="currentUri" value={path} />
          <input type="hidden" name="dashboardUri" value={saveAsDraftUri} />

          <GridRow>
            <GridCol>
              <div id="cmr">
                <MultiChoice
                  meta={{ touched: t(cmrError), error: t(cmrError) }}
                  onChange={this.onChange}
                >
                  <fieldset>
                    <legend className="visually-hidden">
                      {t('sdDoYouHaveaRoadTransportDocumentHeader')}
                    </legend>
                    <SelectRadio
                      id="separateCmrTrue"
                      value={true}
                      name="cmr"
                      inline
                      checked={cmr === 'true'}
                      onChange={this.onChange}
                    >
                      {t('Yes')}
                    </SelectRadio>
                    <SelectRadio
                      id="separateCmrFalse"
                      value={false}
                      name="cmr"
                      inline
                      checked={cmr === 'false'}
                      onChange={this.onChange}
                    >
                      {t('No')}
                    </SelectRadio>
                  </fieldset>
                </MultiChoice>
              </div>
            </GridCol>
          </GridRow>
          <GridRow>
            <SaveAsDraftButton
              formactionUrl={saveAsDraftFormActionUrl}
              onClick={this.onSaveAsDraft}
            />
            <ContinueButton
              type="submit"
              id="continue"
              onClick={this.onContinue}
            >
              {t('commonContinueButtonSaveAndContinueButton')}
            </ContinueButton>
          </GridRow>
        </form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey} />
      </Main>
    );
  };
}

function mapStateToProps(state) {
  return {
    errors: state.errors,
    transport: state.transport,
    getAddedLandingsType: state.landingsType
  };
}

function loadData(store, journey) {
  if (Object.keys(this.queryParams).length !== 0) {
    store.dispatch(dispatchApiCallFailed(JSON.parse(this.queryParams.error)));
  }

  return store.dispatch(getTransportDetails(journey, this.documentNumber))
    .then(() => {
      if (journey === catchCertificateJourney) {
        store.dispatch(getLandingType(this.documentNumber));
      }
    });
}

export const component = connect(
  mapStateToProps,
  {
    addTransportDetails,
    saveTruckCMR,
    getTransportDetails,
    dispatchClearErrors,
    clearTransportDetails,
    getLandingType
  }
)(withRouter(withTranslation() (TruckCMRPage)));

TruckCMRPage.propTypes = {
  transport: PropTypes.object.isRequired,
  errors: PropTypes.object
};

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
