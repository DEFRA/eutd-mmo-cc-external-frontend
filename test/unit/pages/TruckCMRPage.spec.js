import {  mount } from 'enzyme';
import * as React from 'react';
import {component as TruckCMRPage} from '../../../src/client/pages/common/transport/TruckCMRPage';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Router, Route } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import thunk from 'redux-thunk';
import { getLandingType } from '../../../src/client/actions/landingsType.actions';
import { render } from '@testing-library/react';


jest.mock('../../../src/client/actions/landingsType.actions');

describe('TruckCMRPage', () => {

  let wrapper;
  const mockPush = jest.fn();

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({crm: 'true'});
        });
      }}
  })]);

  beforeEach(() => {
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });
    const store = mockStore({ errors: {}, transport: {}, getTransportDetails: {}, saveTruckCMR: {}});

    const props = {
      history : [],
      route   : {
        path : ':documentNumber/transportSelection',
        title       : 'Create a UK catch certificate for exports',
        previousUri : ':documentNumber/transportSelection',
        nextUri     : ':documentNumber/summary',
        saveAsDraftUri: ':documentNumber/create-catch-certificate/catch-certificates',
        landingsEntryUri: ':documentNumber/landings-entry',
        journey: 'catchCertificate' ,
        summaryUri : ':documentNumber/check-your-information',
        truckDetailsUri : ':documentNumber/add-transportation-details-truck',
        progressUri: 'documentNumber/progress'
      }
    };

    window.scrollTo = jest.fn();

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <TruckCMRPage {...props}/>
        </MemoryRouter>
      </Provider>
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render "yes" multi-choice option', () => {
    expect(wrapper.find('input#separateCmrTrue').exists()).toBeTruthy();
  });

  it('should render "no" multi-choice option', () => {
    expect(wrapper.find('input#separateCmrFalse').exists()).toBeTruthy();
  });

  it('should have a for attribute on all input labels', () => {
    expect(wrapper.find('label#label-separateCmrTrue').props()['htmlFor']).toBeDefined();
    expect(wrapper.find('label#label-separateCmrTrue').props()['htmlFor']).toBe('separateCmrTrue');
    expect(wrapper.find('label#label-separateCmrFalse').props()['htmlFor']).toBeDefined();
    expect(wrapper.find('label#label-separateCmrFalse').props()['htmlFor']).toBe('separateCmrFalse');
  });

  it('should handle submit event', () => {
    wrapper.find('form').simulate(
      'submit',
      {preventDefault() {}}
    );
  });

  it('should handle on change events setting CMR to true', () => {
    wrapper.find('input[name="cmr"][id="separateCmrTrue"]').simulate('change', {target: {name: 'cmr',  value: 'true'}});
  });

  it('should handle on change events setting CMR to false', () => {
    wrapper.find('input[name="cmr"][id="separateCmrFalse"]').simulate('change', {target: {name: 'cmr',  value: 'false'}});
  });

  it('should handle save as draft event', () => {
    wrapper.find('button#saveAsDraft').simulate('click');
  });

  it('should have an id on all inputs', () => {
    expect(wrapper.find('input[id="separateCmrTrue"]').exists()).toBeTruthy();
    expect(wrapper.find('input[id="separateCmrFalse"]').exists()).toBeTruthy();
  });

  it('renders Back to Your Progress link with the correct href property', () => {
    expect(wrapper.find('BackToProgressLink')).toBeDefined();
    expect(wrapper.find('BackToProgressLink').find('a').props().href).toBe('/documentNumber/progress');
  });

  it('should take a snapshot of whole page', ()=> {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

  it('should re-direct to forbidden page when unauthorised', async () => {
    const props = {
      history : {
        push: mockPush
      },
      route   : {
        path : ':documentNumber/transportSelection',
        title       : 'Create a UK catch certificate for exports',
        previousUri : ':documentNumber/transportSelection',
        nextUri     : ':documentNumber/summary',
        saveAsDraftUri: ':documentNumber/create-catch-certificate/catch-certificates',
        landingsEntryUri: ':documentNumber/landings-entry',
        journey: 'catchCertificate' ,
        summaryUri : ':documentNumber/check-your-information',
        truckDetailsUri : ':documentNumber/add-transportation-details-truck',
        progressUri: 'documentNumber/progress'
      },
      transport: {
        unauthorised: true
      }
    };

    await new TruckCMRPage.WrappedComponent(props).componentDidUpdate(props);
    expect(mockPush).toHaveBeenCalledWith('/forbidden');
  });
});

describe('TruckCMRPage - with errors', () => {

  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({ crm: 'true' });
        });
      }
    }
  })]);

  beforeEach(() => {
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });
    const store = mockStore({
      errors: {
        cmrError: 'Select yes if you have a road transport document to go with this export',
        errors: [{ targetName: 'cmr', text: 'Select yes if you have a road transport document to go with this export'}]
      },
      transport: {},
      getTransportDetails: {},
      saveTruckCMR: {}
    });

    const props = {
      history: [],
      route: {
        path: ':documentNumber/transportSelection',
        title: 'Create a UK catch certificate for exports',
        previousUri: ':documentNumber/transportSelection',
        nextUri: ':documentNumber/summary',
        saveAsDraftUri: ':documentNumber/create-catch-certificate/catch-certificates',
        landingsEntryUri: ':documentNumber/landings-entry',
        journey: 'catchCertificate',
        summaryUri: ':documentNumber/check-your-information',
        truckDetailsUri: ':documentNumber/add-transportation-details-truck',
        progressUri: 'documentNumber/progress'
      }
    };

    window.scrollTo = jest.fn();

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <TruckCMRPage {...props} />
        </MemoryRouter>
      </Provider>
    );
  });

  it('should render error attributes', () => {
    expect(wrapper.find('ErrorText span').text()).toBe('Select yes if you have a road transport document to go with this export');
  });

  it('should render error summary link with the correct text', () => {
    expect(wrapper.find('ul.error-summary-list li a').text()).toBe('Select yes if you have a road transport document to go with this export');
  });

});

describe('TruckCMRPage - when export is a directLanding', () => {

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const history = createMemoryHistory();
  const mockPush = jest.spyOn(history, 'push');

  beforeEach(() => {
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });

    const store = mockStore({
      errors: {},
      transport: {
        vehicle:'truck'
      },
      getTransportDetails: {},
      landingsType: {
        generatedByContent: false,
        landingsEntryOption: 'directLanding'
      }
    });
    window.scrollTo = jest.fn();

    const props = {
      route: {
        title: 'Create a UK catch certificate for exports',
        previousUri: '/transportSelection',
        nextUri: '/summary',
        showExportDate: true,
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        landingsEntryUri: ':documentNumber/landings-entry',
        journey: 'catchCertificate',
        progressUri: '/create-catch-certificate/:documentNumber/progress'
      },
      match : {
        params : {
          documentNumber: 'GBR-23423-4234234'
        }
      }
    };

    mount(
      <Provider store={store}>
        <Router history={history}>
          <TruckCMRPage {...props}/>
        </Router>
      </Provider>
    );
  });

  afterEach(() => {
    mockPush.mockRestore();
  });

  it('should redirect to exporter dashboard page', () => {
    expect(mockPush).toHaveBeenCalledWith('/create-catch-certificate/catch-certificates');
    expect(mockPush).toHaveBeenCalledTimes(1);
  });
});

describe('TruckCMRPage - when catch cert has no landing type', () => {

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const documentNumber = 'GBR-23423-4234234';
  const history = createMemoryHistory({ initialEntries: [`${documentNumber}/do-you-have-a-road-transport-document`] });
  const mockPush = jest.spyOn(history, 'push');

  const props = {
    route: {
      title: 'Create a UK catch certificate for exports',
      previousUri: '/transportSelection',
      nextUri: '/summary',
      showExportDate: true,
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      landingsEntryUri: ':documentNumber/landings-entry',
      journey: 'catchCertificate',
      progressUri: ':documentNumber/progress'
    }
  };

  beforeEach(() => {
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });

    const store = mockStore({
      errors: {},
      transport: {
        vehicle: 'truck'
      },
      getTransportDetails: {}
    });
    window.scrollTo = jest.fn();

    mount(
      <Provider store={store}>
        <Router history={history}>
          <Route exact path=":documentNumber/do-you-have-a-road-transport-document">
            <TruckCMRPage {...props}/>
          </Route>
        </Router>
      </Provider>
    );
  });

  afterEach(() => {
    mockPush.mockReset();
  });

  it('should redirect to the landing entry page', () => {
    const { landingsEntryUri } = props.route;
    const expectedUrl = landingsEntryUri.replace(':documentNumber', documentNumber);

    expect(mockPush).toHaveBeenCalledWith(expectedUrl);
    expect(mockPush).toHaveBeenCalledTimes(1);
  });
});

describe('TruckCMRPage - when on the storage document journey', () => {
  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const history = createMemoryHistory();
  const mockPush = jest.spyOn(history, 'push');

  beforeEach(() => {
    getLandingType.mockReset();
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });

    const store = mockStore({
      errors: {},
      transport: {
        vehicle: 'truck'
      },
      getTransportDetails: {},
      landingsType: {
        generatedByContent: false,
        landingsEntryOption: 'directLanding'
      }
    });
    window.scrollTo = jest.fn();

    const props = {
      route: {
        title: 'Create a UK catch certificate for exports',
        previousUri: '/transportSelection',
        nextUri: '/summary',
        showExportDate: true,
        saveAsDraftUri: '/create-catch-certificate/catch-certificates',
        landingsEntryUri: ':documentNumber/landings-entry',
        journey: 'storageNotes',
        progressUri: 'documentNumber/progress'
      },
      match : {
        params : {
          documentNumber: 'GBR-23423-4234234'
        }
      }
    };

    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <TruckCMRPage {...props}/>
        </Router>
      </Provider>
    );
  });

  afterEach(() => {
    mockPush.mockReset();
  });

  it('should not call getLandingType', () => {
    expect(getLandingType).not.toHaveBeenCalled();
  });

  it('should not redirect', () => {
    expect(mockPush).not.toHaveBeenCalled();
  });

  it('renders Back to Your Progress link with the correct href property', () => {
    expect(wrapper.find('BackToProgressLink')).toBeDefined();
    expect(wrapper.find('BackToProgressLink').find('a').props().href).toBe('/documentNumber/progress');
  });
});

describe('TruckCMRPage - when tranport details is incomplete', () => {

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const documentNumber = 'GBR-23423-4234234';
  const history = createMemoryHistory({ initialEntries: [`${documentNumber}/do-you-have-a-road-transport-document`] });
  const mockPush = jest.spyOn(history, 'push');

  const props = {
    route: {
      title: 'Create a UK catch certificate for exports',
      previousUri: '/transportSelection',
      nextUri: '/summary',
      showExportDate: true,
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      landingsEntryUri: ':documentNumber/landings-entry',
      journey: 'catchCertificate',
      progressUri: '/create-catch-certificate/:documentNumber/progress'
    },
    match : {
      params : {
        documentNumber: 'GBR-23423-4234234'
      }
    }
  };

  beforeEach(() => {
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });

    const store = mockStore({
      errors: {},
      transport: {},
      getTransportDetails: {},
      getAddedLandingsType: {
        landingsEntryOption: 'manualEntry',
        generatedByContent: true
      }
    });
    window.scrollTo = jest.fn();

    mount(
      <Provider store={store}>
        <Router history={history}>
          <Route exact path=":documentNumber/do-you-have-a-road-transport-document">
            <TruckCMRPage {...props}/>
          </Route>
        </Router>
      </Provider>
    );
  });

  afterEach(() => {
    mockPush.mockReset();
  });

  it('should redirect to the progress page', () => {
    const expectedUrl = props.route.progressUri.replace(':documentNumber', documentNumber);

    expect(mockPush).toHaveBeenCalledWith(expectedUrl);
    expect(mockPush).toHaveBeenCalledTimes(1);
  });
});
