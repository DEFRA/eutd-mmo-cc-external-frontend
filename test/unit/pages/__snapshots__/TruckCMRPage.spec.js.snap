// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TruckCMRPage should take a snapshot of whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="undefined/transportSelection"
          >
            Back
          </a>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            Do you have a road transport document to go with this export?
          </h1>
          <br />
          <span
            class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
          >
            This document is sometimes called a CMR note. Itâ€™s a contract between you and the haulier, and includes details about your export.
          </span>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/transport/truck/cmr"
        method="POST"
      >
        <input
          name="summaryUri"
          type="hidden"
          value=":documentNumber/check-your-information"
        />
        <input
          name="truckDetailsUri"
          type="hidden"
          value=":documentNumber/add-transportation-details-truck"
        />
        <input
          name="journey"
          type="hidden"
          value="catchCertificate"
        />
        <input
          name="vehicle"
          type="hidden"
          value=""
        />
        <input
          name="nationalityOfVehicle"
          type="hidden"
          value=""
        />
        <input
          name="registrationNumber"
          type="hidden"
          value=""
        />
        <input
          name="departurePlace"
          type="hidden"
          value=""
        />
        <input
          name="currentUri"
          type="hidden"
          value=":documentNumber/transportSelection"
        />
        <input
          name="dashboardUri"
          type="hidden"
          value=":documentNumber/create-catch-certificate/catch-certificates"
        />
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-1w80l5h-StyledColumn e4u1u2g0"
          >
            <div
              id="cmr"
            >
              <div
                class="e1vcssft0 css-1phgi8s-StyledFieldset-StyledHoc e4u1u2g0"
              >
                <span
                  class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
                />
                
                <fieldset>
                  <legend
                    class="visually-hidden"
                  >
                    Do you have a road transport document to go with this export?
                  </legend>
                  <label
                    class="css-y8ql23-StyledHoc"
                    for="separateCmrTrue"
                    id="label-separateCmrTrue"
                  >
                    <input
                      class="css-1xt3i3l"
                      id="separateCmrTrue"
                      name="cmr"
                      type="radio"
                      value="true"
                    />
                    <span
                      class="css-1uwxoxa"
                    >
                      Yes
                    </span>
                  </label>
                  <label
                    class="css-y8ql23-StyledHoc"
                    for="separateCmrFalse"
                    id="label-separateCmrFalse"
                  >
                    <input
                      class="css-1xt3i3l"
                      id="separateCmrFalse"
                      name="cmr"
                      type="radio"
                      value="false"
                    />
                    <span
                      class="css-1uwxoxa"
                    >
                      No
                    </span>
                  </label>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/transport/truck/cmr/saveAsDraft"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
            type="submit"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/documentNumber/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
